[Multi-Chain] Add config file

fhenneke opened the pull request:
This PR adds a config file. This configuration is supposed to be extended to other chains in future PRs.

This PR adds a config `AccountingConfig` which consists of multple configurations for different aspects of the accounting.

The config is used as a global config. At the moment it is not explicitly passed to any function.

Next steps:
- Pass config explicitly to functions.
- Make some of the code more general to allow for using it on other chains.
- Add support for other networks. (At the moment I only copied the testing setup on other chains from the original code.)
- Make config more consistent.

There are still hardcoded constants somewhere in the code. One place is when fetching prices in converting token values, via an ID. The id cannot be part of the config due to circular imports. Instead, an address or id string should be passed and the pricing code should turn it into an appropriate type.

### How to review
The first two commits (the names are wrong, I might force push correct ones) add a config and use the config throughout the code. The third commit just removes the old constants file.

One discussion should mostly be about the first commit, i.e. the one adding the configuration classes in `config.py`.
- Should the config be structured differently?
- Do we need special treatment of secrets?

Another discussion would be around how to move forward from here.
- Should we pass the config struct more explicitly? (That is what I did in the more complete code I have.)
- Which parameters are probably still missing for making it work on other chains?

fhenneke committed:
config: fix mypy error

fhenneke committed:
transfer_file: fix lint

fhenneke committed:
delete constants.py

fhenneke committed:
fix unit test

fhenneke committed:
add old default for network

since tests need this variable

fhenneke committed:
make dune key optional for now

this should be removed later. we should
- add special handling of secrets, or
- check this variable at the start of the script

fhenneke committed:
make all secrets optional for now

fhenneke committed:
Merge branch 'main' into add_config

fhenneke committed:
fix merge commit

remove token_list again

fhenneke committed:
incorporate config on sippage into payment script

harisang commented:
@fhenneke can you rebase once the easy PRs are merged, and in case it is ready for review, let's also remove the "draft" label

fhenneke committed:
Merge branch 'main' into add_config

# Conflicts:
#	tests/unit/test_models.py

fhenneke commented:
The things which can already be discussed are:
- The structure of the config class. Is anything obvious missing? Are the abstractions reasonable, i.e., the split into separate configs for rewards, protocol fees, buffer accounting, general payment information, input output config.
- The general design of having a config which can later be passed to all payment functions. That would make it easy to switch between configs without keeping track when environment variables are available. It also makes it easier to check that the config is correct, e.g., that the private key is set if the post transaction command line flag is set.

bram-vdberg commented on src/config.py:
```diff
@@ -0,0 +1,217 @@
+    io_config: IOConfig
+
+
+def get_accounting_config(network: str) -> AccountingConfig:
```
I think it would be easier to maintain if we move this to a classmethod, that way we instantiate the class rather than call a function that returns an object of that class. This would be best if we would want to instantiate the config in different ways (eg. config.from_network() or config.from_env()). 

If there's only ever one way, we could define a post_init that takes a string (or ENUM) to instantiate the config object. [Here is an example of how I set up the config in dune-sync with a post-init method.](https://github.com/cowprotocol/dune-sync/blob/54e22cf2c518820c66824ded0822c1fbddd765ff/src/deploy_prefect/models.py#L56) In this case I think it's better to stick to a classmethod though because the intitialization is a bit more complex.

fhenneke replied:
I moved the logic into a static class method `from_network`.

bram-vdberg commented on src/config.py:
```diff
@@ -0,0 +1,217 @@
+        node_config=NodeConfig(node_url=node_url),
+        reward_config=RewardConfig(
+            reward_token_address=Address("0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB"),
+            batch_reward_cap_upper=12 * 10**15,
```
I think we can make this easier to maintain as well by moving these configs into the RewardConfig class itself and passing the network onto the RewardConfig and setting these properties there based on the value for network.

> reward_config=RewardConfig("mainnet")

fhenneke replied:
I moved all the initializations into their respective classes. Most are `from_network` static method, some are `from_env` static methods.

bram-vdberg commented on src/config.py:
```diff
@@ -0,0 +1,217 @@
+    return result
+
+
+config = get_accounting_config(os.environ.get("NETWORK", "mainnet"))
```
I think it's best to set as an enum, just so we can verify network with `NETWORK.is_mainnet()` rather than `network == "mainnet"`

fhenneke replied:
I added a `Network` enum class.

What is the use case for a function like `is_mainnet`? I am using match clauses for the setup at the moment.

bram-vdberg replied:
I think your approach to use match cases is better actually. I didn't think of doing it like this, using a method like `is_mainnet` is a habit I got into before match cases were introduced to avoid magic strings and make this sort of logic more maintainable. I don't think there's a material difference in functionality, but this looks much more concise.

I think technically match cases are preferred when we want to check multiple different states (such as this where chain can be arbitrum, gnosis, or mainnet). And using a method such as `is_mainnet` would be preferred if there's only one attribute to check with one if statement (such as only performing a certain step on mainnet). I'll make this change on the dune-sync config as well to keep things consistent.

fhenneke committed:
use enum and static class methods for config

this addresses review comments
- uses an enum instead of a string for better checking of network values
- move initialization of configs into individual classes: use from_network and from_env to initialize classes

fhenneke committed:
fix pylint and mypy

mostly by adding comments

fhenneke committed:
Merge branch 'main' into add_config

fhenneke committed:
Merge branch 'main' into add_config

fhenneke committed:
Merge branch 'main' into add_config

# Conflicts:
#	src/fetch/payouts.py

fhenneke committed:
Merge branch 'main' into add_config

# Conflicts:
#	src/fetch/payouts.py
#	tests/unit/test_payouts.py

fhenneke committed:
fix error in test

fhenneke committed:
fix tests a bit more

harisang commented on src/config.py:
```diff
@@ -0,0 +1,294 @@
+
+@dataclass(frozen=True)
+class ProtocolFeeConfig:
+    """Configuration for protocol and parner fees."""
```
```suggestion
    """Configuration for protocol and partner fees."""
```

harisang commented on src/config.py:
```diff
@@ -0,0 +1,294 @@
+
+    protocol_fee_safe: Address
+    partner_fee_cut: float
+    partner_fee_reduced_cut: float
```
This name is not really helpful. Let's add a comment that explains things

harisang commented on src/config.py:
```diff
@@ -0,0 +1,294 @@
+        """Initialize node config for a given network."""
+        # Found this exposed infura key on https://rpc.info/
+        infura_key = os.environ.get("INFURA_KEY", "9aa3d95b3bc440fa88ea12eaa4456161")
+        node_url = f"https://{network}.infura.io/v3/{infura_key}"
```
PR #430 simplifies a bit this part

fhenneke replied:
The new environment variable is used now.

harisang commented on src/config.py:
```diff
@@ -0,0 +1,294 @@
+                safe_url = (
+                    f"https://app.safe.global/{short_name}:{payment_safe_address}"
+                )
+                airdrop_url = (
```
Should we remove this? Seems like an overkill

fhenneke replied:
I can remove it, as it also did not work for me when I tried it some time ago.

Generally, I think we should provide instructions as part of error messages and logging as much as possible.

harisang committed:
Merge branch 'main' into add_config

fhenneke committed:
Merge branch 'main' into add_config

# Conflicts:
#	src/constants.py
#	src/fetch/transfer_file.py

fhenneke committed:
Merge branch 'main' into add_config

fhenneke committed:
fix config import in test_prices

fhenneke committed:
add node url for testing

fhenneke committed:
add empty default node url

harisang committed:
Update src/config.py

harisang commented on src/config.py:
```diff
@@ -0,0 +1,296 @@
+        """Initialize buffer accounting config for a given network."""
+        match network:
+            case Network.MAINNET:
+                return BufferAccountingConfig(include_slippage=True)
```
Does this bypass the flag that is passed as runtime argument?

fhenneke replied:
Currently, the code uses `ignore_slippage_flag or (not config.buffer_accounting_config.include_slippage)`. So if one of them is set to ignore slippage, the slippage part is skipped.

I have a PR almost ready locally to pass the config explicitly everywhere. Then the config is created after parsing of command line arguments and the config object would be set based on the command line flag.

fhenneke committed:
add comments for protocol fee config

# Conflicts:
#	src/config.py

fhenneke committed:
remove airdrop url from config

harisang reviewed (approved):
Makes sense overall, although i admit i haven;t thought too much about its structure.

Definitely nice to save on a lot of imports and random constants showing up in various files.
